api/jsonrpc/eth_endpoint.go

```diff
@@ -138,12 +141,319 @@ func (e *Eth) GetBalance(address string, number string) (interface{}, error) {
+
+	// Fetch the requested header
+	header, err := e.GetBlockHeader(number)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch the world state snapshot
+	s := e.d.minimal.Blockchain.Executor().State()
+	snap, err := s.NewSnapshotAt(header.StateRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	acc, ok := state.NewTxn(s, snap).GetAccount(addr)
+	if !ok {
+		return nil, fmt.Errorf("error getting account state")
+	}
+
+	// Fetch the Storage state snapshot
+	snap, err = s.NewSnapshotAt(acc.Root)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the storage for the passed in location
+	result, ok := snap.Get(index)
+	if !ok {
+		return nil, fmt.Errorf("error getting storage snapshot")
+	}
+
+	return result, nil
+}
+
+// GasPrice returns the average gas price based on the last x blocks
+func (e *Eth) GasPrice() (interface{}, error) {
+
+	// Grab the average gas price and convert it to a hex value
+	avgGasPrice := hex.EncodeBig(e.d.minimal.Blockchain.GetAvgGasPrice())
+
+	return avgGasPrice, nil
+}
+
+// Call executes a smart contract call using the transaction object data
+func (e *Eth) Call(transaction *types.Transaction, number string) (interface{}, error) {
+
+	// Fetch the requested header
+	header, err := e.GetBlockHeader(number)
+	if err != nil {
+		return nil, err
+	}
+
+	transition, err := e.d.minimal.Blockchain.Executor().BeginTxn(header.StateRoot, header)
+	if err != nil {
+		return nil, err
+	}
+
+	// The return value of the execution is saved in the transition (returnValue field)
+	_, _, err = transition.Apply(transaction)
+	if err != nil {
+		return nil, err
+	}
+
+	return transition.ReturnValue(), nil
+}
+
+// GetBlockHeader returns the specific header of the requested block
+// latest, earliest, pending, or a specific block number
+func (e *Eth) GetBlockHeader(number string) (*types.Header, error) {
+
+	blockchainRef := e.d.minimal.Blockchain
+
+	blockRef := "latest" // default value
+
+	if number != "" {
+		blockRef = number
+	}
+
+	switch blockRef {
+	case "latest":
+		header, ok := blockchainRef.Header()
+		if !ok {
+			return nil, fmt.Errorf("Error fetching latest header")
+		}
+
+		return header, nil
+	case "earliest":
+		return nil, fmt.Errorf("Fetching the earliest header is not supported")
+	case "pending":
+		return nil, fmt.Errorf("Fetching the pending header is not supported")
+	default:
+		// Convert the block number from hex to uint64
+		blockNumber := hex.DecodeHexToBig(number)
+
+		header, ok := blockchainRef.GetHeaderByNumber(blockNumber.Uint64())
+		if !ok {
+			return nil, fmt.Errorf("Error fetching block number %d header", blockNumber.Uint64())
+		}
+		return header, nil
+	}
+}
+
+// EstimateGasParams - Optional params used for the EstimateGas call
+type EstimateGasParams struct {
+	transaction *types.Transaction
+	number      string
+}
+
+// EstimateGas estimates the gas needed to execute a transaction
+func (e *Eth) EstimateGas(params EstimateGasParams) (interface{}, error) {
```

Comment:
EstimateGas does not take a block number. Only the transaction object.

Reply:
According to the Ethereum wiki, estimateGas should take in the same params as eth_call, except that they're all optional:

> See eth_call parameters, expect that all properties are optional. If no gas limit is specified geth uses the block gas limit from the pending block as an upper bound. As a result the returned estimate might not be enough to executed the call/transaction when the amount of gas is higher than the pending block gas limit.

> eth_call

Executes a new message call immediately without creating a transaction on the block chain.
Parameters

    Object - The transaction call object

    from: DATA, 20 Bytes - (optional) The address the transaction is sent from.
    to: DATA, 20 Bytes - The address the transaction is directed to.
    gas: QUANTITY - (optional) Integer of the gas provided for the transaction execution. eth_call consumes zero gas, but this parameter may be needed by some executions.
    gasPrice: QUANTITY - (optional) Integer of the gasPrice used for each paid gas
    value: QUANTITY - (optional) Integer of the value sent with this transaction
    data: DATA - (optional) Hash of the method signature and encoded parameters. For details see Ethereum Contract ABI in the Solidity documentation

    QUANTITY|TAG - integer block number, or the string "latest", "earliest" or "pending", see the default block parameter

Reply:
That refers to the transaction object not to the whole set of function params.

Reply:
You are right, estimateGas takes two function params https://github.com/ethereum/go-ethereum/blob/fba5a63afe0993da70896c763b4fdfa953e066ff/internal/ethapi/api.go#L941